### PR TITLE
Use ruff formatter and pre-commit

### PR DESCRIPTION
**Pull request by bouweandela**
## Description

Use [`ruff`](https://docs.astral.sh/ruff/) to format the Python code and [`pre-commit`](https://pre-commit.com/#usage) to manage formatting. Enables [`pre-commit.ci`](https://results.pre-commit.ci/latest/github/ESMValGroup/ESMValCore/main) to check that code has been formatted correctly.

These new tools replace `yapf`, `isort` and `flake8` (`pyflakes` and `pycodestyle`). I have dropped `docformatter` from the pre-commit hooks because it makes changes to docstrings that do not always look good (e.g. it will happily cut a sentence into two pieces to make the first half of the sentence fit on the first line of the docstring).

Please install the pre-commit hooks by going to the directory where you have checked out ESMValCore and running
```
pre-commit install
```
To format your code (and check for pycodestyle, pyflakes, and mypy issues) run:
```
pre-commit run -a
```
This will also automatically be done (on changed files only) whenever you commit your changes.

See https://github.com/ESMValGroup/ESMValTool/discussions/2161

Link to documentation: https://esmvaltool--2524.org.readthedocs.build/projects/ESMValCore/en/2524/contributing.html#code-quality

Git branch upgrade instructions (once this has been merged)
------------------------------------------------
To upgrade your existing git branches and open pull requests to the new formatting once this pull request has been merged into the `main` branch, the following procedure is recommended to minimize merge conflicts:

Step 1: reformat changed Python files with ruff, this is needed to minimize merge conflicts in the next step
```bash
git checkout your-branch
pip install ruff
ruff format --config line-length=79 $(git diff --name-only origin/main... '*.py')
git add -u
git commit -m 'Apply ruff formatting to changed files' -n
```
Step 2: merge the `main` branch into your branch
```bash
git pull origin main --no-ff
# edit code to address merge conflicts, use `git add path/to/file.py` to mark as solved and then run:
git commit -n
```
Step 3: start using pre-commit to format and check your code
```bash
pre-commit install
pre-commit run -a
# edit code to fix any issues that ruff did not solve automatically
git add -u
git commit -m 'Fix linter issues'
git push
```
To test these instructions before this has been merged, you can replace `main` with `ruff-format` (this branch) in the commands above.
***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful (except the code coverage threshold because some reformatted code is untested and Codacy which thinks existing issues are new because the code they occur is has been reformatted).

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by valeriupredoi** [comment]
seems it doesn't like Python 3.10 both Linux and OSX, lemme have a look at that, bud

**Comment by bouweandela** [comment]
Thanks! I think codespell needed an extra dependency to read the configuration file in toml format on Python 3.10. Added in fe3ff7b.

**Comment by valeriupredoi** [comment]
> Thanks! I think codespell needed an extra dependency to read the configuration file in toml format on Python 3.10. Added in [fe3ff7b](https://github.com/ESMValGroup/ESMValCore/commit/fe3ff7b098cd98376f0fe5324ffc0efefcb3062c).

:black_cat: :smiley:

**Comment by valeriupredoi** [comment]
are we going to accept ruffly 92% coverage as the new project coverage?

**Comment by bouweandela** [comment]
> are we going to accept ruffly 92% coverage as the new project coverage?

My idea was to assume that reformatting is a safe operation and does not introduce bugs.. writing test for all the lines missing coverage seems a considerable amount of work.

**Comment by valeriupredoi** [comment]
> > are we going to accept ruffly 92% coverage as the new project coverage?
> 
> My idea was to assume that reformatting is a safe operation and does not introduce bugs.. writing tests for all the lines missing coverage seems a considerable amount of work.

oh for sure! But I am confused as to why the coverage drops vs the original codebase, since this PR only reformats code, and doesn't introduce any hefty functionality :confused:

File: .circleci/config.yml
Comment (valeriupredoi) [inline .circleci/config.yml:31]: are we not running flake8 since pre-commit runs flake8-compatible tests?

File: .github/workflows/create-condalock-file.yml
Comment (valeriupredoi) [inline .github/workflows/create-condalock-file.yml:77]:
```suggestion
```

**Comment by bouweandela** [comment]
It's only the diff coverage that is partial, if you click the little arrow near the bottom of the Codecov comment where it says "Additional details and impacted files" it says that overall coverage is unchanged.

File: .github/workflows/create-condalock-file.yml
Comment (valeriupredoi) [inline .github/workflows/create-condalock-file.yml:75]: no need to run either flake8 nor pre-commit on this one since the lockfile doesn't introduce any code changes (my historical bad, sorry); if you want to be very careful maybe just run a call to eg `pre-commit --version`

File: .yamllint
Comment (valeriupredoi) [inline .yamllint:8]: from 79 to 120 now??

**Comment by valeriupredoi** [comment]
> It's only the diff coverage that is partial, if you click the little arrow near the bottom of the Codecov comment where it says "Additional details and impacted files" it says that overall coverage is unchanged.

ah always found these confusing af, thanks, bud! Phew :face_exhaling:

**Review by valeriupredoi — APPROVED** [review]
really hope you didn't have to do all these changes manually, bud :rofl: Cheers muchly! We have to start thinking how we do that in ESMValTool - gonna be messy. What we doing about pull requests after we merge this - they'll be conflicting like mad I imagine

File: .circleci/config.yml
Comment (bouweandela) [reply .circleci/config.yml:31]: Yes: I enabled only those linting rules for `ruff` that are exactly the same as what flake8 provides (see [here](https://docs.astral.sh/ruff/faq/#how-does-ruffs-linter-compare-to-flake8) for more info). Therefore ruff replaces flake8. In the future we could consider enabling more linter rules.

File: .yamllint
Comment (bouweandela) [reply .yamllint:8]: Mostly from an error to a warning because for yaml files the formatting can really become a lot worse if you shorten the lines. We already had this limit in ESMValTool (for those long dataset entries in recipes), so that's why I copied it here. Happy to change it to 79 if you think that's more consistent though.

Comment (valeriupredoi) [reply .yamllint:8]: mrh up to you. I thought the linter error changed, and now we'll see 120 char lines everywhere :grin:

File: .circleci/config.yml
Comment (bouweandela) [reply .circleci/config.yml:31]: And indeed those are already run on pre-commit.ci (also mypy), so I didn't see the point in running them again on CircleCI.

File: .yamllint
Comment (bouweandela) [reply .yamllint:8]: https://github.com/ESMValGroup/ESMValTool/blob/main/.yamllint

File: .github/workflows/create-condalock-file.yml
Comment (valeriupredoi) [reply .github/workflows/create-condalock-file.yml:75]: yes, good idea! No need to slow down the tests

File: .yamllint
Comment (valeriupredoi) [reply .yamllint:8]: OK, proves out I am talking nonsense :laughing:

File: .github/workflows/create-condalock-file.yml
Comment (valeriupredoi) [reply .github/workflows/create-condalock-file.yml:75]: is that a Python `venv` or a conda one? We may want to tell it to use mamba if it's a conda one

Comment (bouweandela) [reply .github/workflows/create-condalock-file.yml:75]: It says is works without any system dependencies for Python, but if I run the pre-commit hooks in a Python 3.10 conda environment that does seem to be picked up [as shown by the failing tests with the GitHub action for Python 3.10](https://github.com/ESMValGroup/ESMValCore/actions/runs/10921655486/job/30314252157), so I'm not quite sure if I understand how that works.

Comment (valeriupredoi) [reply .github/workflows/create-condalock-file.yml:75]: OK pip is fine, like other such env-in-env applications. Was scared it's a conda fanboi :grin:

File: .github/workflows/install-from-condalock-file.yml
Comment (bouweandela) [inline .github/workflows/install-from-condalock-file.yml:56]:
```suggestion
```

File: .github/workflows/create-condalock-file.yml
Comment (bouweandela) [reply .github/workflows/create-condalock-file.yml]: Done in 93cfe33

Comment (valeriupredoi) [reply .github/workflows/create-condalock-file.yml]: yay :beer:

Comment (valeriupredoi) [reply .github/workflows/create-condalock-file.yml]: ahha - starting to make sense of this pre-commit thingie: you know it installs all stuff in ` $HOME/.cache/pre-commit` dir right?

Comment (valeriupredoi) [reply .github/workflows/create-condalock-file.yml]: OK so what it does is it installs all those deps listed in the yaml file separately, each in its own venv, so they don't kill each other in a single common venv, my `.cache/pre-commit` looks like this:
```
 (esmvaltool) valeriu@valeriu-PORTEGE-Z30-C:~/ESMValTool$ ls -la ~/.cache/pre-commit/
total 72
drwxrwxr-x 11 valeriu valeriu  4096 Sep 18 16:19 .
drwx------ 16 valeriu valeriu  4096 Sep 18 16:16 ..
-rw-------  1 valeriu valeriu 20480 Sep 18 16:16 db.db
-rw-rw-r--  1 valeriu valeriu     0 Sep 18 16:13 .lock
-rw-rw-r--  1 valeriu valeriu  2284 Sep 18 16:17 pre-commit.log
-rw-rw-r--  1 valeriu valeriu   109 Sep 18 16:13 README
drwx------  4 valeriu valeriu  4096 Sep 18 16:28 repo4ep1qzk5
drwx------ 10 valeriu valeriu  4096 Sep 18 16:16 repo4uqei5t0
drwx------ 13 valeriu valeriu  4096 Sep 18 16:16 repo7jrwosxf
drwx------  7 valeriu valeriu  4096 Sep 18 16:16 repo7xmez37c
drwx------ 10 valeriu valeriu  4096 Sep 18 16:16 repoa4zz9bxd
drwx------  8 valeriu valeriu  4096 Sep 18 16:16 repoadrq8rrg
drwx------ 12 valeriu valeriu  4096 Sep 18 16:17 repoo3owvyb7
drwx------  9 valeriu valeriu  4096 Sep 18 16:16 repote4lws0p
drwx------  4 valeriu valeriu  4096 Sep 18 16:16 repotlhn3o_9
```
each of those repoxxx are seprate venvs - by the looks of it those deps need be in the master Python environment (in our case Miniforge etc) so pre-commit can grab the dependencies and put them in those little single-occupancy venvs, that's why it needed `codespell` in the Py3.10 envs in those tests

Comment (valeriupredoi) [reply .github/workflows/create-condalock-file.yml]: ah there you go, explained by none other than Anthony Sottile the pre-commit developer https://stackoverflow.com/questions/66174906/pre-commit-not-finding-python-packages

**Review by valeriupredoi — APPROVED** [review]
last changes are ruffling no feathers :feather:

File: .prospector.yml
Comment (schlunma) [inline .prospector.yml:18]: Is there some kind of reference that you could add as a comment here?

**Review by schlunma — APPROVED** [review]
I love it!! ❤️ This makes our code so much cleaner.

Two comments:

1. Does pre-commit.ci make any changes to the code? Or does it only check the code?
2. Should we consider tackling some of the Codacy issues here? Looks like most of them are very easy to fix. Just a suggestion, don't want to put extra work on you.

Thanks so much!!

File: .prospector.yml
Comment (bouweandela) [reply .prospector.yml:21]: Done in 7b0f8cd

**Comment by bouweandela** [comment]
> Does pre-commit.ci make any changes to the code? Or does it only check the code?

It is possible to configure pre-commit.ci such that it will automatically fix issues and push a commit to your branch. However, I disabled it for now because I am concerned that this may cause trouble for some of our contributors if they forget to install the pre-commit hooks and then pre-commit.ci commits to their branch, they continue editing and add some more commits, try to `git push` them, and find that they cannot because the branch on GitHub conflicts with their local branch. Of course, they could `git push --force`, but they may not be familiar enough with git to realize that.

> Should we consider tackling some of the Codacy issues here? Looks like most of them are very easy to fix. Just a suggestion, don't want to put extra work on you.

I would prefer not to make any more changes here because there are so many already that it is almost impossible to review. `ruff` can also automatically fix many issues, so if we enable more `ruff` rules over time, many of these issues will probably be automatically fixed.

**Comment by valeriupredoi** [comment]
> > Does pre-commit.ci make any changes to the code? Or does it only check the code?
> 
> It is possible to configure pre-commit.ci such that it will automatically fix issues and push a commit to your branch. However, I disabled it for now because I am concerned that this may cause trouble for some of our contributors if they forget to install the pre-commit hooks and then pre-commit.ci commits to their branch, they continue editing and add some more commits, try to `git push` them, and find that they cannot because the branch on GitHub conflicts with their local branch. Of course, they could `git push --force`, but they may not be familiar enough with git to realize that.
> 
> > Should we consider tackling some of the Codacy issues here? Looks like most of them are very easy to fix. Just a suggestion, don't want to put extra work on you.
> 
> I would prefer not to make any more changes here because there are so many already that it is almost impossible to review. `ruff` can also automatically fix many issues, so if we enable more `ruff` rules over time, many of these issues will probably be automatically fixed.

fully support these two from Bouwe. We may, however, need to turn on auto-pre-commit commits done by Circle (as Bouwe had it turned on for a bit until I barked at him to turn it off), for some of the older PRs that will need pre-commit be run on them, and the contributor may not be doing it or be weary/scared/not know how to do it (I reckon those aren't many, since most Core devs know their stuff). That, in the case of merge `main` (after this PR gets merged) will result in too many a conflict, that is

**Comment by schlunma** [comment]
However, I disabled it for now because I am concerned that this may cause trouble for some of our contributors if they forget to install the pre-commit hooks and then pre-commit.ci commits to their branch, they continue editing and add some more commits, try to `git push` them, and find that they cannot because the branch on GitHub conflicts with their local branch.

Perfect, this is what I wanted to hear!

Will look at https://github.com/ESMValGroup/ESMValCore/pull/2529 now to fix the tests, then we can hopefully merge this soon.

**Comment by bouweandela** [comment]
I fixed the tests for now by pinning the pandas version to not 2.2.*.

**Comment by valeriupredoi** [comment]
> I fixed the tests for now by pinning the pandas version to not 2.2.*.

Peasant solution, more peasant than my peasant rounding in #2529 🐑

**Pull request merged**
